### PR TITLE
Add support for eslint-plugin-typescript

### PR DESCRIPTION
**Pull request by j-f1**
@lydell Can you take a look and see if I did anything wrong?

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -632,6 +638,46 @@ Example ESLint configuration:
 }
 ```
 
+### [typescript/member-delimiter-style]
```
Comment (lydell) [inline README.md]: When is this rule useful to have enabled at the same time as using Prettier?

File: README.md [hunk 2]
```diff
@@ -632,6 +638,46 @@ Example ESLint configuration:
...
+        "multiline": {
+          "delimiter": "none"
+        }
+      }
+    ]
+  }
+}
+```
+
+### [typescript/type-annotation-spacing]
```
Comment (lydell) [inline README.md]: When is this rule useful to have enabled at the same time as using Prettier?

File: README.md (hunk 1 above)
Comment (j-f1) [reply README.md]: I added it since it follows a similar rationale to `vue/html-self-closing` — it works fine with Prettier when configured as specified.

File: README.md (hunk 2 above)
Comment (j-f1) [reply README.md]: ⏫

File: README.md (hunk 1 above)
Comment (lydell) [reply README.md]: Ok, but that's not what this config tries to do. There are plenty of rules that work just fine if configured correctly. But even then they don't provide any value. They just take time to execute and run the risk of conflicting in the future.

Comment (j-f1) [reply README.md]: Removed 🔥

File: test-lint/typescript.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+"use strict";
+
+// Prettier wants semicolons as delimiters, but the "typescript" rule added in
+// .eslintrc.base.js wants commas.
+interface Foo {
+  a: Foo;
+  b: Bar;
+}
+
+function add(a: number, b: number): number {
```
Comment (lydell) [inline test-lint/typescript.ts]: Is this function needed?

File: test/validators.test.js [hunk 4]
```diff
@@ -1,21 +1,39 @@
...
 });
 
-test("lines-around-comment", () => {
-  expect(validators["lines-around-comment"]([])).toBe(false);
-  expect(
-    validators["lines-around-comment"]([
+function rule(name, { valid, invalid }) {
+  test(name, () => {
+    const validator = validators[name];
+    if (valid) {
```
Comment (lydell) [inline test/validators.test.js]: Are `valid` and `invalid` ever absent?

**Comment by j-f1** [comment]
Removed 🔥

File: .eslintrc.base.js [hunk 5]
```diff
@@ -40,8 +42,26 @@ module.exports = {
...
+    "typescript/no-var-requires": "off",
+    "typescript/no-use-before-define": "off",
+    "typescript/type-annotation-spacing": [
+      "error",
+      { before: true, after: false }
+    ]
   },
+  overrides: [
+    {
+      files: ["*.ts"],
```
Comment (BPScott) [inline .eslintrc.base.js]: Driveby comment:  Don't forget `*.tsx` files

File: .eslintrc.base.js [hunk 6]
```diff
@@ -20,6 +21,7 @@ module.exports = {
     parser: "babel-eslint",
     ecmaVersion: 2018,
     sourceType: "script",
+    loggerFn: () => {},
```
Comment (lydell) [inline .eslintrc.base.js]: What does this do?

Comment (lydell) [reply .eslintrc.base.js]: Never mind, it's needed to avoid this error:

```
 FAIL  test/lint-verify-fail.test.js
  ● test-lint/ causes errors without eslint-config-prettier › encountered a declaration exception

    SyntaxError: Unexpected token = in JSON at position 0
        at JSON.parse (<anonymous>)

      16 |     { encoding: "utf8", env: getEnv() }
      17 |   );
    > 18 |   const output = JSON.parse(result.stdout);
         |                       ^
      19 | 
      20 |   test("every test-lint/ file must cause an error", () => {
      21 |     expect(output.length).toBe(ruleFiles.length);

      at Suite.parse (test/lint-verify-fail.test.js:18:23)
      at Object.describe (test/lint-verify-fail.test.js:12:1)

```

File: .eslintrc.base.js [hunk 7]
```diff
@@ -42,6 +44,17 @@ module.exports = {
...
     "standard/computed-property-even-spacing": ["error", "even"]
   },
+  overrides: [
+    {
+      files: ["*.ts", "*.tsx"],
+      parserOptions: { parser: "eslint-plugin-typescript/parser" },
+      rules: {
+        // Force a conflict with Prettier in test-lint/typescript.js.
+        // This is included in "plugin:typescript/recommended".
+        "typescript/indent": "error"
```
[the remark this answers is not in the document]
Comment (lydell) [reply .eslintrc.base.js]: Yes!

[the remark this answers is not in the document]
Comment (j-f1) [reply .eslintrc.base.js]: Is https://github.com/prettier/eslint-config-prettier/pull/68/commits/27df642081cc09c0874a9cc840c90fa94f65f766 what you were looking for?

[the remark this answers is not in the document]
Comment (lydell) [reply .eslintrc.base.js]: I tried that myself, but ESLint does not seem to allow `extends` in `overrides`. Travis fails:

```
Error: ESLint configuration in /home/travis/build/prettier/eslint-config-prettier/.eslintrc.base.js is invalid:
	- Unexpected top-level property "overrides[0].extends".
```

[the remark this answers is not in the document]
Comment (lydell) [reply .eslintrc.base.js]: I think we should roll back to my last commit and merge.

If you find out a way to use ESLint both for JS and TS in the same project, I'd be interested to know. I guess it works if you don't use extend.

[the remark this answers is not in the document]
Comment (j-f1) [reply .eslintrc.base.js]: Rolled back.

**Comment by IlyaSemenov** [comment]
Can this be published to npm now? Thanks. (By the way did you consider https://github.com/semantic-release/semantic-release ?)

**Comment by lydell** [comment]
I'll make a release later today.
